File: lab3/src/main/java/lab3/functionality/Player.java
package lab3.functionality;

public class Player extends Actionable {

    private Health health;
    private Location curLocation;
    private boolean completedQuest;

    public Player(String name, Location startLocation) {
        super(name, p -> String.format("Состояние %s\nЗдоровье: %s", p.getName(), p.getHealth().getName()));
        setHealth(Health.HEALTHY);
        setLocation(startLocation);
        setCompletedQuest(false);
    }

    public boolean isCompletedQuest() {
        return completedQuest;
    }

    public void setCompletedQuest(boolean completedQuest) {
        this.completedQuest = completedQuest;
    }

    public Location getLocation() {
        return curLocation;
    }

    public String setLocation(Location location) {
        this.curLocation = location;
        return location.enter(this);
    }

    public Health getHealth() {
        return health;
    }

    public void setHealth(Health health) {
        this.health = health;
    }

    public void damage() {
        setHealth(getHealth().next());
    }

    public void regen() {
        setHealth(getHealth().previous());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Player player)) return false;

        return getName().equals(player.getName());
    }

    @Override
    public int hashCode() {
        return getName() != null ? getName().hashCode() : 0;
    }
}

File: lab4/src/main/kotlin/lab4/functionality/Player.kt
package lab4.functionality

class Player(
    name: String,
    var location: Location
) : Actionable(
    name,
    { p -> "Состояние ${p.name}\nЗдоровье: ${p.health.state}" }
) {
    var health: Health = Health.HEALTHY
    var completedQuest: Boolean = false

    fun enterLocation(location: Location){
        this.location = location
        location.enter(this)
    }

    fun damage(){
        health = health.next()
    }

    fun heal() {
        health = health.previous()
    }
}


